### PR TITLE
[WIP] Use yaml as the only configuration file format

### PR DESCRIPTION
**Pull request by LaPetiteSouris**
This PR is to support YAML as the only configuration format for the project. Eventually TOML and JSON can be used without any code modification (unittests are needed though). Reasons behind the change:

- YAML is great for configuration and widely accepted

- Maybe we need to move away from current config format (NATS style), and use something more universal (TOML/YAML)

Changes have been made:

- Instead of parsing log file manually, I use package called [viper](https://github.com/spf13/viper). The package itself support parsing `yaml` and `toml` .  
- Replace `.conf` file with `.yaml` file. Technically speaking, as `viper` supports `yaml`, `toml` , `json` out of the box, it should work with any of these file formats. However I tested with only `yaml` for now. The question is : Should we support all of these formats ? Or should we make only one standard configuration file format as `yaml` ?
- Delete the code base dealing with old `conf` file format.  Personally I am in favor of supporting only one single file format as configuration, as the behavior is more predictable, and the effort to convert from old `.conf` file to `yaml` file is relatively light, I deleted old code base instead of supporting legacy. How do you think about that?
- Add new tests  and change old tests to adapt to the modifications.
- Separate `log` section and `stream` section in configuration file. Currently `log.level` and `log.retention` creates confusion with `log` section. The current `log` section is renamed to `stream` to explicitly reflect its function.

Finally, I am new to Golang , so apologize if something is too "smelly" in the code.

### CODE REVIEW

**Comment by tylertreat** [comment]
Thanks for the PR, will review shortly. FYI there is an [issue](https://github.com/liftbridge-io/liftbridge/issues/71) for tracking this work. I was planning on going with TOML as the config format. I'm nervous about supporting lots of different config types (JSON, TOML, YAML, etc.) because I think it gets confusing from a usability and documentation perspective versus having a single canonical format.

**Comment by LaPetiteSouris** [comment]
Thanks. I will try to fix the builds soon when I can.
If `TOML` is the choice, then it should be fine as well since `viper` support it out of the box without modification. However, if that the case then I will need to modify the sample configuration file to `toml` format and double check the tests.
What is the advantage of toml over yaml btw ?

**Comment by tylertreat** [comment]
> What is the advantage of toml over yaml btw ?

Mostly preference I think. :) 

Let's leave it as yaml for now. I think yaml is the most common, especially with k8s, so that might be the more appropriate choice.

File: server/config.go [hunk 1]
```diff
@@ -203,81 +208,110 @@ func GetLogLevel(level string) (uint32, error) {
...
+		config.DataDir = v.GetString("data.dir")
+	}
+
+	if v.IsSet("batch.max.messages") {
+		config.BatchMaxMessages = v.GetInt("batch.max.messages")
 	}
 
+	if v.IsSet("batch.wait.time") {
+		waitTime := v.GetString("batch.wait.time")
+		dur, err := time.ParseDuration(waitTime)
```
Comment (tylertreat) [inline server/config.go]: Can we use `v.GetDuration` instead?

File: server/config.go [hunk 2]
```diff
@@ -203,81 +208,110 @@ func GetLogLevel(level string) (uint32, error) {
...
+		if err != nil {
+			return nil, err
+		}
+		config.BatchWaitTime = dur
+	}
+
+	if v.IsSet("metadata.cache.max.age") {
+		maxAge := v.GetString("metadata.cache.max.age")
+		dur, err := time.ParseDuration(maxAge)
```
Comment (tylertreat) [inline server/config.go]: `v.GetDuration`?

File: server/config.go [hunk 3]
```diff
@@ -288,122 +322,140 @@ func NewConfig(configFile string) (*Config, error) { // nolint: gocyclo
...
+		config.Log.RetentionMaxBytes = v.GetInt64("log.retention.max.bytes")
+	}
+
+	if v.IsSet("log.retention.max.messages") {
+		config.Log.RetentionMaxMessages = v.GetInt64("log.retention.max.messages")
+	}
+
+	if v.IsSet("log.retention.max.age") {
+		durMaxAgeStr := v.GetString("log.retention.max.age")
+		dur, err := time.ParseDuration(durMaxAgeStr)
```
Comment (tylertreat) [inline server/config.go]: `v.GetDuration`?

File: server/config.go [hunk 4]
```diff
@@ -288,122 +322,140 @@ func NewConfig(configFile string) (*Config, error) { // nolint: gocyclo
...
+		if err != nil {
+			return err
+		}
+		config.Log.RetentionMaxAge = dur
+	}
+
+	if v.IsSet("log.cleaner.interval") {
+		cleanerIntervalStr := v.GetString("log.cleaner.interval")
+		dur, err := time.ParseDuration(cleanerIntervalStr)
```
Comment (tylertreat) [inline server/config.go]: `v.GetDuration`?

File: server/config.go [hunk 5]
```diff
@@ -288,122 +322,140 @@ func NewConfig(configFile string) (*Config, error) { // nolint: gocyclo
...
+		config.Log.CleanerInterval = dur
+	}
+
+	if v.IsSet("log.segment.max.bytes") {
+		config.Log.SegmentMaxBytes = v.GetInt64("log.segment.max.bytes")
 	}
+
+	if v.IsSet("log.roll.time") {
+		rollTimeStr := v.GetString("log.roll.time")
+		dur, err := time.ParseDuration(rollTimeStr)
```
Comment (tylertreat) [inline server/config.go]: `v.GetDuration`?

File: server/config.go [hunk 6]
```diff
@@ -288,122 +322,140 @@ func NewConfig(configFile string) (*Config, error) { // nolint: gocyclo
...
+	}
+	if v.IsSet("clustering.raft.bootstrap.peers") {
+		config.Clustering.RaftBootstrapPeers = v.GetStringSlice("clustering.raft.bootstrap.peers")
+	}
+
+	if v.IsSet("clustering.raft.logging") {
+		config.Clustering.RaftLogging = v.GetBool("clustering.raft.logging")
+	}
+	if v.IsSet("clustering.replica.max.lag.time") {
+		dur, err := time.ParseDuration(v.GetString("clustering.replica.max.lag.time"))
```
Comment (tylertreat) [inline server/config.go]: `v.GetDuration`?

File: server/config.go [hunk 7]
```diff
@@ -288,122 +322,140 @@ func NewConfig(configFile string) (*Config, error) { // nolint: gocyclo
...
+		if err != nil {
+			return err
 		}
+		config.Clustering.ReplicaMaxLagTime = dur
+	}
+
+	if v.IsSet("clustering.replica.max.leader.timeout") {
+		dur, err := time.ParseDuration(v.GetString("clustering.replica.max.leader.timeout"))
```
Comment (tylertreat) [inline server/config.go]: `v.GetDuration`?

File: server/config.go [hunk 8]
```diff
@@ -288,122 +322,140 @@ func NewConfig(configFile string) (*Config, error) { // nolint: gocyclo
...
+		if err != nil {
+			return err
+		}
+		config.Clustering.ReplicaMaxLeaderTimeout = dur
+	}
+
+	if v.IsSet("clustering.replica.max.idle.wait") {
+		dur, err := time.ParseDuration(v.GetString("clustering.replica.max.idle.wait"))
```
Comment (tylertreat) [inline server/config.go]: `v.GetDuration`?

File: server/config.go [hunk 9]
```diff
@@ -288,122 +322,140 @@ func NewConfig(configFile string) (*Config, error) { // nolint: gocyclo
...
+		if err != nil {
+			return err
+		}
+		config.Clustering.ReplicaMaxIdleWait = dur
+	}
+
+	if v.IsSet("clustering.replica.fetch.timeout") {
+		dur, err := time.ParseDuration(v.GetString("clustering.replica.fetch.timeout"))
```
Comment (tylertreat) [inline server/config.go]: `v.GetDuration`?

File: server/config.go [hunk 10]
```diff
@@ -118,6 +117,12 @@ type Config struct {
 	Clustering          ClusteringConfig
 }
 
+// new Viper to parse configuration file
+func newViper() *viper.Viper {
+	v := viper.New()
+	return v
+}
```
[the remark this answers is not in the document]
Comment (LaPetiteSouris) [reply server/config.go]: At first I though we may need additional configuration for `viper`, but indeed it makes sense just to call `viper.New()` directly.
Fixed

File: server/config.go [hunk 11]
```diff
@@ -203,81 +208,110 @@ func GetLogLevel(level string) (uint32, error) {
...
-		return nil, err
-	}
+	v := newViper()
+	// Expect a config.yaml file in the destination
+	v.SetConfigFile(configFile)
 
+	if err := v.ReadInConfig(); err != nil {
+		if _, ok := err.(viper.ConfigFileNotFoundError); ok {
+			// return default config
+			return config, nil
```
[the remark this answers is not in the document]
Comment (LaPetiteSouris) [reply server/config.go]: Make sense. Fixed

File: server/config.go [hunk 12]
```diff
@@ -203,81 +208,110 @@ func GetLogLevel(level string) (uint32, error) {
...
+		}
+
+		return nil, fmt.Errorf("Error on loading config: %v", err)
```
[the remark this answers is not in the document]
Comment (LaPetiteSouris) [reply server/config.go]: Fixed

File: server/config.go [hunk 13]
```diff
@@ -415,16 +467,17 @@ type HostPort struct {
 }
 
 // parseListen will parse the `listen` option containing the host and port.
-func parseListen(v interface{}) (*HostPort, error) {
+func parseListen(v *viper.Viper) (*HostPort, error) {
```
[the remark this answers is not in the document]
Comment (LaPetiteSouris) [reply server/config.go]: Yes we could do that. Initialize I did it that way , but just some thoughts:
- As `viper` gives a lot of flexibility in parsing config, I think we can pass `viper` pointer and let the function deal with the config as it wishes
- I think using `viper.GetString(myConfKey)` looks better than parsing it manually with type checking as in the old way.
- Thus, `parseListen`, `parseLog` and `parseClustering` use the same method to get the config, which is given the `viper` pointer and let them deal with it as they want

What is your opinion about that ?

File: documentation/configuration.md [hunk 14]
```diff
@@ -49,49 +49,45 @@ GLOBAL OPTIONS:
...
-    replica.max.lag.time: "20s"
-}
+---
+host: 0.0.0.0
+port: '5050'
+data:
+  dir: "/foo"
+# Log configuration
+log:
+  level: 'debug'
```
[the remark this answers is not in the document]
Comment (LaPetiteSouris) [reply documentation/configuration.md]: shall we go with `streaming` ? Indeed I am confused about this `log` section

File: server/config.go (hunk 6 above)
Comment (LaPetiteSouris) [reply server/config.go]: fixed

File: server/config.go (hunk 7 above)
Comment (LaPetiteSouris) [reply server/config.go]: fixed.

File: server/config.go (hunk 8 above)
Comment (LaPetiteSouris) [reply server/config.go]: fixed

File: server/config.go (hunk 9 above)
Comment (LaPetiteSouris) [reply server/config.go]: fixed

File: documentation/configuration.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (tylertreat) [reply documentation/configuration.md]: There are currently two logging related configs: `log.level` and `log.recovery`. My thought is these should become `logging.level` and `logging.recovery`.

The `log` section is used to configure settings for message streams (i.e. "logs"). I'm thinking this section should become `streams`, e.g. `streams.retention.max.bytes`, `streams.segment.max.bytes`, `streams.compact`, etc.

[the remark this answers is not in the document]
Comment (LaPetiteSouris) [reply documentation/configuration.md]: This makes sense. Thus, I did:

- Use `streams` instead  of ` log` in configuration
- The `log` section now represents logging related configs `log.level` and `log.recovery`.
- The `Config` struct is also refactored to ensure coherence in the modification.

[the remark this answers is not in the document]
Comment (tylertreat) [reply documentation/configuration.md]: It looks like the changes are properly covered so don't worry about coverage. 80% is just target coverage level.

**Pull request merged**